nbextensions/usability/toc2/main.js

```diff
@@ -0,0 +1,422 @@
+             'toc_cell':false,
+             'toc_window_display':false}
+    var threshold = cfg['toc_threshold']
+    var toc_cell=cfg['toc_cell'];
+    var number_sections = cfg['toc_number_sections'];
+
+
+   function read_config() {  // read after nb is loaded
+     // create config object to load parameters
+    var base_url = utils.get_body_data("baseUrl");
+    var config = new configmod.ConfigSection('notebook', {base_url: base_url});
+
+    // update params with any specified in the server's config file
+    var update_params = function() {
+        for (var key in cfg) {
+            if (config.data.hasOwnProperty(key)){
+                cfg[key] = config.data[key];
+}
+        }
+        IPython.notebook.metadata.toc = cfg; //save in present nb metadata (then can be modified per document)
+        threshold = cfg['toc_threshold'];
+        toc_cell=cfg['toc_cell'];
+        number_sections = cfg['toc_number_sections'];
+        //$('#toc-wrapper').css('display',cfg['toc-wrapper_display']) //ensure display is done as noted in config
+        $('#toc-wrapper').css('display',cfg['toc_window_display'] ? 'block' : 'none') //ensure display is done as noted in config
+    };
+
+    // config may be specified at system level or at document level.
+    if (IPython.notebook.metadata.toc !== undefined){ //configuration saved in nb
+        console.log("config stored in nb")
+	cfg = IPython.notebook.metadata.toc;
+        threshold = cfg['toc_threshold'];
+        toc_cell=cfg['toc_cell'];
+        number_sections = cfg['toc_number_sections'];
+	}
+    else {
+        console.log("config stored in system")
+	config.load();
+        config.loaded.then(function() {update_params() })
+    } 
+    config_loaded = true;
+}
+//.....................global variables....
+
+    var rendering_toc_cell = false; 
+    var config_loaded = false;
+    var extension_initialized=false;
+
+//......... utilitary functions............    
+
+  var make_link = function (h) {
+    var a = $("<a/>");
+    a.attr("href", '#' + h.attr('id'));
+    // get the text *excluding* the link text, whatever it may be
+    var hclone = h.clone();
+    hclone.children().last().remove(); // remove only the last child 
+    a.html(hclone.html());
+    a.on('click',function(){setTimeout(function(){console.log('clicked'); $.ajax()}, 100) }) //workaround for  https://github.com/jupyter/notebook/issues/699
+                                                                                        //as suggested by @jhamrick
+    return a;
+  };
+
+  var ol_depth = function (element) {
+    // get depth of nested ol
+    var d = 0;
+    while (element.prop("tagName").toLowerCase() == 'ol') {
+      d += 1;
+      element = element.parent();
+    }
+    return d;
+  };
+  
+  var create_toc_div = function () {
+    var toc_wrapper = $('<div id="toc-wrapper"/>')
+    .append(
+      $("<div/>")
+      .addClass("header")
+      .text("Contents ")
+      .click( function(){
```

Comment:
One thing I noticed is, that dragging the TOC to move it also collapses it.
Maybe it would be better to move the click function down to line 115 so the TOC will only collapse when you click on the '-' ?